nncf/experimental/onnx/graph/onnx_graph.py

```diff
@@ -202,24 +205,83 @@ def get_weight_tensor(self, node: onnx.NodeProto) -> Tuple[str, np.ndarray]:
         :param node: Node, in which the weight tensor finally applied.
         :return: Weight tensor name and its value.
         """
-        node_name = node.name
-        node_inputs = self.get_node_edge_names(node_name)['input']
-        weight_input = node_inputs[1]
         try:
+            weight_port_id = self.get_weight_tensor_port_id(node)
+            weight_input = self.get_node_edge_names(node.name)['input'][weight_port_id]
             return self.get_initializer(weight_input).name, self.get_initializers_value(weight_input)
-        except RuntimeError as e:
-            node = self.get_nodes_by_output(weight_input)[0]
-            metatype = ONNX_OPERATION_METATYPES.get_operator_metatype_by_op_name(node.op_type)
-            if metatype == ONNXIdentityMetatype:
-                tensor_name = self.get_initializer(node.input[0]).name
-                return tensor_name, self.get_initializers_value(tensor_name)
-            if metatype == ONNXReshapeMetatype:
-                tensor_name = node.output[0]
-                shape = self.get_initializers_value(node.input[1])
-                tensor_value = self.get_initializers_value(node.input[0])
-                reshaped_tensor_value = tensor_value.reshape(shape)
-                return tensor_name, reshaped_tensor_value
-            raise RuntimeError('Could not find the weight value of the node') from e
+        except RuntimeError as exc:
+            weight_port_id = self.get_weight_tensor_port_id(node)
+            parent_node_on_weight_port = self.get_parents(node)[weight_port_id]
+            nodes = deque([parent_node_on_weight_port])
+            while nodes:
+                try:
+                    current_node = nodes.popleft()
+                    node_parents = self.get_parents(current_node)
+                    nodes.extendleft(node_parents)
+                    metatype = ONNX_OPERATION_METATYPES.get_operator_metatype_by_op_name(current_node.op_type)
+                    if metatype in [ONNXIdentityMetatype, ONNXQuantizeLinearMetatype]:
+                        return self._get_tensor_from_zero_input(current_node)
+                    if metatype == ONNXReshapeMetatype:
+                        return self._get_weight_tensor_with_reshape(current_node)
+                    weight_port_id = self.get_weight_tensor_port_id(current_node)
+                    if weight_port_id is None:
+                        # The node does not have weight
+                        continue
+                    weight_input = self.get_node_edge_names(current_node.name)['input'][weight_port_id]
+                    return self.get_initializer(weight_input).name, self.get_initializers_value(weight_input)
+                except RuntimeError:
+                    continue
+            raise RuntimeError('Could not find the weight value of the node') from exc
+
+    def _get_param_from_weight_definitions(self, node: onnx.NodeProto, parameter: str) -> int:
```

Comment:
Not good - you are basically wrapping around the `getattr` usage and using `getattr` is generally wrong save for special cases. If you really need to get something from the `OpWeightDef`, and the `OpWeightDef` may not be there, why not `_get_weight_definitions` and then just access the `OpWeightDef` statically defined members via `.` access?

Reply:
This func was added to remove copy/paste code for the getting attributes of `OpWeightDef`. There are three of them:
```
get_weight_port_id
get_weight_channel_axis
get_bias_tensor_port_id
```

Reply:
How does using the approach I described lead to copy/paste code?

Reply:
Updated. Please take a look